File: theseus_acc/a2ui.py
"""A2UI EvidenceCockpit scene builders.

Converts an ACCReport into a JSON-serializable scene that the browser
extension's A2UI catalog renders without running a model. Model-generated
explanations can be layered on top with a validator-checked
``ModelExplanationPanel`` while preserving ACC scores.

This module never calls a model, never reaches the network, and never depends
on private Theseus code. It is published as part of the public theseus_acc
package.
"""

from __future__ import annotations

from typing import Any

from .algorithm import ACCReport, ClaimACC
from .rules import ACC_V2_VERSION
from .schemas import (
    ALWAYS_PRESENT_COMPONENTS,
    CALIBRATION_SOURCES,
    COMPONENT_CALIBRATION_BADGE,
    COMPONENT_CLAIM_CARD,
    COMPONENT_CONTRADICTION_PANEL,
    COMPONENT_MODEL_EXPLANATION_PANEL,
    COMPONENT_NEXT_CHECKS,
    COMPONENT_PENALTY_LIST,
    COMPONENT_RULE_CHECKLIST,
    COMPONENT_SOURCE_COLLAPSE_PANEL,
    COMPONENT_TRAIT_RADAR,
    REQUIRED_PROPS_BY_COMPONENT,
    SCENE_NAME,
    SCENE_REQUIRED_FIELDS,
    SCENE_VERSION,
)


def _component(component_type: str, claim_id: str, props: dict[str, Any]) -> dict[str, Any]:
    """Build a single component dict with stable id."""
    return {
        'type': component_type,
        'id': f'{component_type}.{claim_id}',
        'props': props,
    }


def _claim_card_props(claim_id: str, claim_text: str, entry: ClaimACC) -> dict[str, Any]:
    return {
        'claim_id': claim_id,
        'claim_text': claim_text,
        'claim_state': entry.claim_state,
        'support_strength': round(float(entry.support_strength), 6),
        'epistemic_risk': round(float(entry.epistemic_risk), 6),
        'verification_gap': entry.verification_gap,
    }


def _source_collapse_props(entry: ClaimACC) -> dict[str, Any] | None:
    diag = entry.diagnostics or {}
    visible = int(diag.get('visible_source_count', 0) or 0)
    if visible < 2:
        return None
    canonical = int(diag.get('canonical_origin_count', 0) or 0)
    ratio = float(diag.get('source_collapse_ratio', 0.0) or 0.0)
    if ratio >= 0.5:
        warning = (
            'Many citations trace back to fewer canonical origins. '
            'Find an independent primary source.'
        )
    elif ratio >= 0.25:
        warning = 'Some citations share canonical origins. Worth verifying independence.'
    else:
        warning = 'Citations appear to come from independent canonical origins.'
    return {
        'visible_source_count': visible,
        'canonical_origin_count': canonical,
        'source_collapse_ratio': round(ratio, 6),
        'warning': warning,
    }


def _trait_radar_props(entry: ClaimACC, weights: dict[str, float]) -> dict[str, Any]:
    return {
        'traits': {k: round(float(v), 6) for k, v in entry.traits.items()},
        'weights': {k: round(float(v), 6) for k, v in weights.items()},
    }


def _rule_checklist_props(entry: ClaimACC) -> dict[str, Any]:
    return {
        'rules': [
            {
                'id': rule['id'],
                'passed': bool(rule['passed']),
                'value': round(float(rule['value']), 6),
                'threshold': round(float(rule['threshold']), 6),
                'reason': rule['reason'],
            }
            for rule in entry.rules
        ],
    }


def _penalty_list_props(entry: ClaimACC) -> dict[str, Any] | None:
    if not entry.penalties:
        return None
    return {
        'penalties': [
            {
                'id': pen['id'],
                'severity': round(float(pen['severity']), 6),
                'weight': round(float(pen['weight']), 6),
                'impact': round(float(pen['impact']), 6),
                'reason': pen['reason'],
            }
            for pen in entry.penalties
        ],
        'penalty_total': round(float(entry.penalty_total), 6),
    }


def _contradiction_panel_props(entry: ClaimACC) -> dict[str, Any] | None:
    diag = entry.diagnostics or {}
    contradictions = int(diag.get('contradiction_count', 0) or 0)
    if contradictions <= 0:
        return None
    return {'contradiction_count': contradictions}


def _next_checks_props(entry: ClaimACC) -> dict[str, Any]:
    return {
        'actions': [
            {
                'id': action['id'],
                'priority': action['priority'],
                'reason': action['reason'],
            }
            for action in entry.actions
        ],
    }


def _calibration_badge_props(entry: ClaimACC, threshold: float) -> dict[str, Any]:
    return {
        'source': 'deterministic',
        'score': round(float(entry.acc_score), 6),
        'threshold': round(float(threshold), 6),
        'version': entry.version,
    }


def _model_explanation_props(value: Any) -> dict[str, Any] | None:
    if isinstance(value, str):
        summary = value.strip()
        citations: list[str] = []
    elif isinstance(value, dict):
        summary = str(value.get('summary', '')).strip()
        raw_citations = value.get('citations', [])
        citations = []
        if isinstance(raw_citations, list):
            citations = [str(c).strip() for c in raw_citations if str(c).strip()]
    else:
        return None

    if not summary:
        return None

    props: dict[str, Any] = {'summary': summary}
    if citations:
        props['citations'] = citations
    return props


def build_evidence_scene(
    report: ACCReport,
    *,
    claim_texts: dict[str, str] | None = None,
    model_explanations: dict[str, Any] | None = None,
    summary: str | None = None,
    calibration_source: str | None = None,
) -> dict[str, Any]:
    """Build an EvidenceCockpit A2UI scene from an ACCReport.

    Args:
        report: ACCReport produced by compute_acc.
        claim_texts: Optional override mapping {claim_id: text}. When absent
            the scene uses an empty string for claim_text; the browser
            extension's claim graph builder fills this in before rendering.
        model_explanations: Optional mapping {claim_id: summary or payload}.
            When provided, ModelExplanationPanel components are emitted and
            CalibrationBadge.source defaults to ``model-adjusted``.
        summary: Optional scene-level one-paragraph summary.
        calibration_source: Explicit CalibrationBadge source. Defaults to
            ``model-adjusted`` when explanations are present; otherwise
            ``deterministic``.

    Returns:
        A JSON-serializable dict matching the EvidenceCockpit schema.
        The scene always contains a ClaimCard, TraitRadar, RuleChecklist,
        NextChecks, and CalibrationBadge per scored claim. Conditional
        components (SourceCollapsePanel, ContradictionPanel, PenaltyList)
        appear only when their underlying signal is non-trivial.
    """
    claim_texts = claim_texts or {}
    clean_model_explanations = {
        claim_id: props
        for claim_id, value in (model_explanations or {}).items()
        if (props := _model_explanation_props(value)) is not None
    }
    source = calibration_source or (
        'model-adjusted' if clean_model_explanations else 'deterministic'
    )
    if source not in CALIBRATION_SOURCES:
        raise ValueError(f'invalid calibration_source: {source}')
    components: list[dict[str, Any]] = []

    for claim_id, entry in report.scores.items():
        text = claim_texts.get(claim_id, '')

        components.append(_component(
            COMPONENT_CLAIM_CARD,
            claim_id,
            _claim_card_props(claim_id, text, entry),
        ))

        collapse_props = _source_collapse_props(entry)
        if collapse_props is not None:
            components.append(_component(
                COMPONENT_SOURCE_COLLAPSE_PANEL,
                claim_id,
                collapse_props,
            ))

        components.append(_component(
            COMPONENT_TRAIT_RADAR,
            claim_id,
            _trait_radar_props(entry, report.weights),
        ))

        components.append(_component(
            COMPONENT_RULE_CHECKLIST,
            claim_id,
            _rule_checklist_props(entry),
        ))

        penalty_props = _penalty_list_props(entry)
        if penalty_props is not None:
            components.append(_component(
                COMPONENT_PENALTY_LIST,
                claim_id,
                penalty_props,
            ))

        contradiction_props = _contradiction_panel_props(entry)
        if contradiction_props is not None:
            components.append(_component(
                COMPONENT_CONTRADICTION_PANEL,
                claim_id,
                contradiction_props,
            ))

        components.append(_component(
            COMPONENT_NEXT_CHECKS,
            claim_id,
            _next_checks_props(entry),
        ))

        explanation_props = clean_model_explanations.get(claim_id)
        if explanation_props is not None:
            components.append(_component(
                COMPONENT_MODEL_EXPLANATION_PANEL,
                claim_id,
                explanation_props,
            ))

        components.append(_component(
            COMPONENT_CALIBRATION_BADGE,
            claim_id,
            {
                **_calibration_badge_props(entry, report.threshold),
                'source': source,
            },
        ))

    return {
        'scene': SCENE_NAME,
        'version': SCENE_VERSION,
        'acc_version': ACC_V2_VERSION,
        'claim_count': len(report.scores),
        'threshold': float(report.threshold),
        'components': components,
        'summary': (summary or '').strip(),
    }


def build_model_adjusted_evidence_scene(
    report: ACCReport,
    *,
    claim_texts: dict[str, str] | None = None,
    model_explanations: dict[str, Any],
    summary: str | None = None,
) -> dict[str, Any]:
    """Build a validator-safe model-adjusted scene.

    The model may supply explanatory text only. ACC scores, thresholds,
    rules, penalties, and actions still come from the deterministic report.
    """
    return build_evidence_scene(
        report,
        claim_texts=claim_texts,
        model_explanations=model_explanations,
        summary=summary,
        calibration_source='model-adjusted',
    )


def validate_evidence_scene(scene: Any) -> list[str]:
    """Return a list of human-readable validation errors.

    An empty list means the scene matches the EvidenceCockpit schema.
    Used by the browser extension to reject malformed Gemma output and
    by the test suite to catch drops in the deterministic builder.
    """
    errors: list[str] = []

    if not isinstance(scene, dict):
        return ['scene is not a JSON object']

    for field in SCENE_REQUIRED_FIELDS:
        if field not in scene:
            errors.append(f'scene missing required field: {field}')

    if scene.get('scene') != SCENE_NAME:
        errors.append(f'scene name must be "{SCENE_NAME}"')

    components = scene.get('components')
    if not isinstance(components, list):
        errors.append('components must be a list')
        return errors

    claim_card_seen: set[str] = set()
    components_by_claim: dict[str, set[str]] = {}
    calibration_source_by_claim: dict[str, str] = {}
    model_explanation_claims: set[str] = set()

    for index, component in enumerate(components):
        if not isinstance(component, dict):
            errors.append(f'component[{index}] is not a JSON object')
            continue
        ctype = component.get('type')
        cid = component.get('id')
        props = component.get('props')
        if not isinstance(ctype, str):
            errors.append(f'component[{index}] missing type')
            continue
        if not isinstance(cid, str):
            errors.append(f'component[{index}] missing id')
        if not isinstance(props, dict):
            errors.append(f'component[{index}] ({ctype}) missing props')
            continue

        required = REQUIRED_PROPS_BY_COMPONENT.get(ctype)
        if required is None:
            errors.append(f'component[{index}] unknown type "{ctype}"')
            continue

        for prop_name in required:
            if prop_name not in props:
                errors.append(f'component[{index}] ({ctype}) missing prop: {prop_name}')

        if ctype == COMPONENT_CALIBRATION_BADGE:
            source = props.get('source')
            if source not in CALIBRATION_SOURCES:
                errors.append(
                    f'component[{index}] CalibrationBadge has invalid source "{source}"'
                )
            if isinstance(cid, str) and '.' in cid and isinstance(source, str):
                calibration_source_by_claim[cid.split('.', 1)[1]] = source

        if ctype == COMPONENT_CLAIM_CARD:
            claim_id = props.get('claim_id')
            if isinstance(claim_id, str):
                claim_card_seen.add(claim_id)
        else:
            # Component id format is "<type>.<claim_id>"; recover claim id.
            if isinstance(cid, str) and '.' in cid:
                claim_id = cid.split('.', 1)[1]
                bucket = components_by_claim.setdefault(claim_id, set())
                bucket.add(ctype)
                if ctype == COMPONENT_MODEL_EXPLANATION_PANEL:
                    model_explanation_claims.add(claim_id)

    claim_count = scene.get('claim_count')
    if isinstance(claim_count, int) and claim_count != len(claim_card_seen):
        errors.append(
            f'claim_count {claim_count} does not match number of ClaimCards {len(claim_card_seen)}'
        )

    for claim_id in claim_card_seen:
        present = components_by_claim.get(claim_id, set())
        for component_type in ALWAYS_PRESENT_COMPONENTS:
            if component_type == COMPONENT_CLAIM_CARD:
                continue
            if component_type not in present:
                errors.append(
                    f'claim "{claim_id}" missing required component: {component_type}'
                )
        if (
            claim_id in model_explanation_claims
            and calibration_source_by_claim.get(claim_id) != 'model-adjusted'
        ):
            errors.append(
                f'claim "{claim_id}" has ModelExplanationPanel but CalibrationBadge source is not model-adjusted'
            )

    return errors

File: browser-extension/src/inference/a2ui.js
/**
 * Strict EvidenceCockpit A2UI scene validator and builder for the browser
 * extension side. Mirrors theseus_acc/a2ui.py.
 *
 * The validator is the security boundary that rejects Gemma output that
 * drops props, alters ACC scores, or invents claim ids. The builder
 * converts the existing scoring.js article-level + per-claim output into
 * the same scene shape the Python builder emits, optionally layering
 * model-written explanation panels on top of deterministic ACC scores.
 *
 * Shape contract is documented in docs/a2ui-scene-schema.md.
 */

import { ALGORITHM_VERSION } from "../shared/config.js";
import {
  ALWAYS_PRESENT_COMPONENTS,
  CALIBRATION_SOURCES,
  COMPONENT_CALIBRATION_BADGE,
  COMPONENT_CLAIM_CARD,
  COMPONENT_CONTRADICTION_PANEL,
  COMPONENT_MODEL_EXPLANATION_PANEL,
  COMPONENT_NEXT_CHECKS,
  COMPONENT_PENALTY_LIST,
  COMPONENT_RULE_CHECKLIST,
  COMPONENT_SOURCE_COLLAPSE_PANEL,
  COMPONENT_TRAIT_RADAR,
  REQUIRED_PROPS_BY_COMPONENT,
  SCENE_NAME,
  SCENE_REQUIRED_FIELDS,
  SCENE_VERSION,
} from "./schemas.js";

function round6(n) {
  return Number(Number(n).toFixed(6));
}

function clamp01(n) {
  return round6(Math.max(0, Math.min(1, Number(n))));
}

const CANONICAL_TRAIT_KEYS = [
  "root_depth",
  "source_independence",
  "support_ratio",
  "claim_specificity",
  "temporal_spread",
  "evidence_volume",
  "falsifiability",
  "rhetorical_pressure",
  "source_quality",
  "contradiction_load",
  "citation_chain_collapse",
];

function firstFinite(...values) {
  for (const value of values) {
    if (value != null && Number.isFinite(Number(value))) {
      return Number(value);
    }
  }
  return null;
}

function withCanonicalTraits(features) {
  const out = { ...(features || {}) };
  const external = firstFinite(out.external_support_ratio);
  const consensus = firstFinite(out.consensus_alignment);
  if (out.support_ratio == null) {
    if (external != null && consensus != null) {
      out.support_ratio = clamp01(Math.min(external, consensus));
    } else if (external != null || consensus != null) {
      out.support_ratio = clamp01(external ?? consensus);
    }
  }
  const aliases = {
    falsifiability: firstFinite(out.falsifiability, out.claim_falsifiability),
    rhetorical_pressure: firstFinite(out.rhetorical_pressure, out.rhetorical_red_flags),
    source_quality: firstFinite(out.source_quality, out.source_tier),
    contradiction_load: firstFinite(out.contradiction_load, out.consensus_alignment, 1),
    citation_chain_collapse: firstFinite(out.citation_chain_collapse, out.citation_chain_closure),
  };
  for (const [key, value] of Object.entries(aliases)) {
    if (out[key] == null && value != null) {
      out[key] = clamp01(value);
    }
  }
  return out;
}

function makeComponent(type, claimId, props) {
  return {
    type,
    id: `${type}.${claimId}`,
    props,
  };
}

function normalizeModelExplanations(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }
  return value;
}

function modelExplanationProps(value) {
  let summary = "";
  let citations = [];

  if (typeof value === "string") {
    summary = value.trim();
  } else if (value && typeof value === "object" && !Array.isArray(value)) {
    summary = String(value.summary ?? "").trim();
    if (Array.isArray(value.citations)) {
      citations = value.citations.map((item) => String(item).trim()).filter(Boolean);
    }
  } else {
    return null;
  }

  if (!summary) {
    return null;
  }

  const props = { summary };
  if (citations.length) {
    props.citations = citations;
  }
  return props;
}

function hasModelExplanation(modelExplanations) {
  return Object.values(modelExplanations).some((value) => modelExplanationProps(value));
}

/**
 * Build an EvidenceCockpit scene from scoring.js output.
 *
 * @param {object} scoreResult - The object returned by scoreText() in
 *   inference/scoring.js. Must include `claims`, `features`, `rules`,
 *   `penalties`, `actions`, `weight_profile_used`, etc.
 * @param {object} [options]
 * @param {string} [options.calibrationSource] - Defaults to "model-adjusted"
 *   when model explanations are present; otherwise "deterministic".
 * @param {object} [options.modelExplanations] - Mapping of claim id to
 *   explanation summary or `{summary, citations}` payload.
 * @param {string} [options.summary] - Optional scene-level summary.
 * @param {number} [options.threshold] - Suspect threshold; defaults to 0.55.
 * @returns {object} A JSON-serializable EvidenceCockpit scene.
 */
export function buildEvidenceScene(scoreResult, options = {}) {
  const modelExplanations = normalizeModelExplanations(
    options.modelExplanations ?? scoreResult?.model_explanations ?? scoreResult?.modelExplanations,
  );
  const calibrationSource = options.calibrationSource
    || (hasModelExplanation(modelExplanations) ? "model-adjusted" : "deterministic");
  const threshold = typeof options.threshold === "number" ? options.threshold : 0.55;
  const summary = String(options.summary ?? scoreResult?.summary ?? "").trim();

  if (!CALIBRATION_SOURCES.includes(calibrationSource)) {
    throw new Error(`invalid calibrationSource: ${calibrationSource}`);
  }

  const components = [];
  const claims = Array.isArray(scoreResult?.claims) ? scoreResult.claims : [];
  const overallTraits = scoreResult?.features ? { ...scoreResult.features } : {};
  // Resolved per-key weights are written by callers when they want them
  // surfaced inside TraitRadar.props.weights. Default to an empty object
  // so the validator's required-prop check passes without forcing a
  // circular import on scoring.js.
  const weightsByProfile = {};

  // The browser extension scores per-claim; emit one set of components
  // per claim. The article-level "overall" score is exposed through the
  // CalibrationBadge of the synthetic _article_overall claim if no real
  // claims were extracted.
  if (!claims.length) {
    const claimId = "article_overall";
    components.push(...buildComponentsForRecord({
      claimId,
      claimText: "",
      record: {
        score: scoreResult?.overall_score ?? 0,
        verdict: scoreResult?.verdict ?? "unreliable",
        rules: scoreResult?.rules ?? [],
        penalties: scoreResult?.penalties ?? [],
        actions: scoreResult?.actions ?? [],
        feature_breakdown: overallTraits,
        linear_score: scoreResult?.linear_score ?? 0,
        geometric_core: scoreResult?.geometric_core ?? 0,
        penalty_total: scoreResult?.penalty_total ?? 0,
      },
      weights: weightsByProfile,
      calibrationSource,
      modelExplanation: modelExplanationProps(modelExplanations[claimId]),
      threshold,
    }));
  } else {
    for (const claim of claims) {
      components.push(...buildComponentsForRecord({
        claimId: claim.id,
        claimText: claim.text || "",
        record: claim,
        weights: weightsByProfile,
        calibrationSource,
        modelExplanation: modelExplanationProps(modelExplanations[claim.id]),
        threshold,
      }));
    }
  }

  return {
    scene: SCENE_NAME,
    version: SCENE_VERSION,
    acc_version: ALGORITHM_VERSION,
    claim_count: claims.length || 1,
    threshold,
    summary,
    components,
  };
}

export function buildModelAdjustedEvidenceScene(scoreResult, modelExplanations, options = {}) {
  return buildEvidenceScene(scoreResult, {
    ...options,
    modelExplanations,
    calibrationSource: "model-adjusted",
  });
}

function buildComponentsForRecord({
  claimId,
  claimText,
  record,
  weights,
  calibrationSource,
  modelExplanation,
  threshold,
}) {
  const list = [];
  const features = record.feature_breakdown || {};

  list.push(
    makeComponent(COMPONENT_CLAIM_CARD, claimId, {
      claim_id: claimId,
      claim_text: claimText,
      claim_state: classifyClaimState(record, threshold),
      support_strength: round6(supportStrengthFromRecord(record)),
      epistemic_risk: round6(epistemicRiskFromRecord(record)),
      verification_gap: verificationGapFromRecord(record),
    }),
  );

  const collapseProps = sourceCollapsePropsFromRecord(record, features);
  if (collapseProps) {
    list.push(makeComponent(COMPONENT_SOURCE_COLLAPSE_PANEL, claimId, collapseProps));
  }

  list.push(
    makeComponent(COMPONENT_TRAIT_RADAR, claimId, {
      traits: roundFeatures(features),
      weights: weights || {},
    }),
  );

  list.push(
    makeComponent(COMPONENT_RULE_CHECKLIST, claimId, {
      rules: (record.rules || []).map(roundRule),
    }),
  );

  if ((record.penalties || []).length) {
    list.push(
      makeComponent(COMPONENT_PENALTY_LIST, claimId, {
        penalties: record.penalties.map(roundPenalty),
        penalty_total: round6(record.penalty_total || 0),
      }),
    );
  }

  const contradictionCount = inferContradictionCount(features);
  if (contradictionCount > 0) {
    list.push(
      makeComponent(COMPONENT_CONTRADICTION_PANEL, claimId, {
        contradiction_count: contradictionCount,
      }),
    );
  }

  list.push(
    makeComponent(COMPONENT_NEXT_CHECKS, claimId, {
      actions: (record.actions || []).map((action) => ({
        id: action.id,
        priority: action.priority,
        reason: action.reason,
      })),
    }),
  );

  if (modelExplanation) {
    list.push(makeComponent(COMPONENT_MODEL_EXPLANATION_PANEL, claimId, modelExplanation));
  }

  list.push(
    makeComponent(COMPONENT_CALIBRATION_BADGE, claimId, {
      source: calibrationSource,
      score: round6(record.score || 0),
      threshold: round6(threshold),
      version: ALGORITHM_VERSION,
    }),
  );

  return list;
}

function supportStrengthFromRecord(record) {
  if (Number.isFinite(Number(record.support_strength))) {
    return clamp01(record.support_strength);
  }
  const linear = Number(record.linear_score || 0);
  const core = Number(record.geometric_core || 0);
  return clamp01(0.65 * linear + 0.35 * core);
}

function epistemicRiskFromRecord(record) {
  if (Number.isFinite(Number(record.epistemic_risk))) {
    return clamp01(record.epistemic_risk);
  }
  // Mirrors the Python formula from compute_acc; the JS scorer does not
  // produce all five components, so use the available proxies:
  //   penalty_total / 0.45                        weight 0.40
  //   1 - source_independence (article)           weight 0.30
  //   1 - root_depth (article)                    weight 0.20
  //   1 - evidence_volume (article)               weight 0.10
  const features = withCanonicalTraits(record.feature_breakdown || {});
  const penalty = Math.min(1, Number(record.penalty_total || 0) / 0.45);
  const indep = 1 - clamp01(features.source_independence || 0);
  const rootless = 1 - clamp01(features.root_depth || 0);
  const thin = 1 - clamp01(features.evidence_volume || 0);
  return clamp01(0.4 * penalty + 0.3 * indep + 0.2 * rootless + 0.1 * thin);
}

function verificationGapFromRecord(record) {
  if (typeof record.verification_gap === "string") {
    return record.verification_gap;
  }
  const features = withCanonicalTraits(record.feature_breakdown || {});
  const independence = Number(features.source_independence ?? 1);
  const root = Number(features.root_depth ?? 0);
  const contradiction = Number(features.contradiction_load ?? features.consensus_alignment ?? 1);
  const evidence = Number(features.evidence_volume ?? 0);
  const specificity = Number(features.claim_specificity ?? 0);
  const temporal = Number(features.temporal_spread ?? 1);

  if (record.score == null && (record.rules || []).length === 0) {
    return "No supporting evidence has been linked yet.";
  }
  if (independence < 0.35 && (record.feature_breakdown?.evidence_volume ?? 0) > 0) {
    return "Multiple citations trace back to fewer canonical origins. Find an independent primary source.";
  }
  if (root < 0.25) {
    return "Needs a verified or reviewed primary root.";
  }
  if (contradiction < 0.5) {
    return "Contradiction pressure outweighs support. Adjudicate before promoting.";
  }
  if (evidence < 0.3) {
    return "Direct support volume is too low. Gather more evidence.";
  }
  if (specificity < 0.25) {
    return "Claim is too vague to adjudicate cleanly. Sharpen with concrete anchors.";
  }
  if (temporal < 0.15 && evidence >= 0.5) {
    return "Evidence collapses into a narrow time window. Broaden temporal sampling.";
  }
  return "";
}

function classifyClaimState(record, threshold) {
  if (typeof record.claim_state === "string" && record.claim_state) {
    return record.claim_state;
  }
  const features = withCanonicalTraits(record.feature_breakdown || {});
  const score = Number(record.score || 0);
  const independence = Number(features.source_independence ?? 1);
  const evidence = Number(features.evidence_volume ?? 0);
  const root = Number(features.root_depth ?? 0);
  const closure = Number(features.citation_chain_collapse ?? features.citation_chain_closure ?? 1);
  const contradiction = Number(features.contradiction_load ?? features.consensus_alignment ?? 1);
  const specificity = Number(features.claim_specificity ?? 0);
  const branchCount = (record.rules || []).find((r) => r.id === "requires_independent_sources")?.value ?? 0;

  if (score >= Math.max(0.65, threshold) && independence >= 0.5 && root >= 0.25 && contradiction >= 0.5) {
    return "well_supported";
  }
  if (closure < 0.5 && independence < 0.5) {
    return "source_collapsed";
  }
  if (contradiction < 0.5) {
    return "contradicted";
  }
  if (evidence < 0.3 || branchCount === 0) {
    return "under_evidenced";
  }
  if (root < 0.25) {
    return "rootless";
  }
  if (specificity < 0.25) {
    return "vague";
  }
  if (score < threshold) {
    return "suspect";
  }
  return "unresolved";
}

function sourceCollapsePropsFromRecord(record, features) {
  if (record.diagnostics && Number(record.diagnostics.visible_source_count || 0) >= 2) {
    const collapseRatio = clamp01(record.diagnostics.source_collapse_ratio ?? 0);
    if (collapseRatio <= 0) {
      return null;
    }
    return {
      visible_source_count: Number(record.diagnostics.visible_source_count || 0),
      canonical_origin_count: Number(record.diagnostics.canonical_origin_count || 0),
      source_collapse_ratio: round6(collapseRatio),
      warning:
        collapseRatio >= 0.5
          ? "Many citations trace back to fewer canonical origins. Find an independent primary source."
          : "Some citations share canonical origins. Worth verifying independence.",
    };
  }
  // The browser scoring path doesn't track canonical_origin counts; fall
  // back to citation_chain_closure as a proxy: closure < 0.5 indicates
  // collapse pressure. Skip the panel when no signal is present.
  const normalized = withCanonicalTraits(features);
  const closure = Number(normalized.citation_chain_collapse ?? normalized.citation_chain_closure ?? 1);
  if (closure >= 0.7) {
    return null;
  }
  // Approximate visible/canonical from independence and closure.
  // independence captures top-share; lower independence = higher collapse.
  const independence = clamp01(normalized.source_independence ?? 1);
  const visible = Math.max(2, Math.round(record.feature_breakdown?.evidence_volume * 8) || 2);
  const collapseRatio = clamp01(1 - closure);
  const canonical = Math.max(1, Math.round(visible * (1 - collapseRatio)));
  let warning;
  if (collapseRatio >= 0.5) {
    warning =
      "Many citations trace back to fewer canonical origins. Find an independent primary source.";
  } else if (collapseRatio >= 0.25) {
    warning = "Some citations share canonical origins. Worth verifying independence.";
  } else {
    warning = "Citations appear to come from independent canonical origins.";
  }
  return {
    visible_source_count: visible,
    canonical_origin_count: canonical,
    source_collapse_ratio: round6(collapseRatio),
    warning,
    independence_proxy: round6(independence),
  };
}

function inferContradictionCount(features) {
  const normalized = withCanonicalTraits(features);
  const contradiction = Number(normalized?.contradiction_load ?? 1);
  if (contradiction >= 0.5) {
    return 0;
  }
  return Math.max(1, Math.round((1 - contradiction) * 3));
}

function roundFeatures(features) {
  const out = {};
  const normalized = withCanonicalTraits(features);
  for (const k of CANONICAL_TRAIT_KEYS) {
    const v = normalized[k];
    if (v == null) {
      continue;
    }
    out[k] = round6(v);
  }
  return out;
}

function roundRule(rule) {
  return {
    id: rule.id,
    passed: Boolean(rule.passed),
    value: rule.value == null ? null : round6(rule.value),
    threshold: round6(rule.threshold),
    reason: rule.reason,
  };
}

function roundPenalty(penalty) {
  return {
    id: penalty.id,
    severity: round6(penalty.severity),
    weight: round6(penalty.weight),
    impact: round6(penalty.impact),
    reason: penalty.reason,
  };
}

/**
 * Strict validator. Returns a list of human-readable error strings.
 * Empty list means the scene is valid.
 *
 * @param {unknown} scene
 * @returns {string[]}
 */
export function validateEvidenceScene(scene) {
  const errors = [];
  if (!scene || typeof scene !== "object" || Array.isArray(scene)) {
    return ["scene is not a JSON object"];
  }

  for (const field of SCENE_REQUIRED_FIELDS) {
    if (!(field in scene)) {
      errors.push(`scene missing required field: ${field}`);
    }
  }

  if (scene.scene !== SCENE_NAME) {
    errors.push(`scene name must be "${SCENE_NAME}"`);
  }

  if (!Array.isArray(scene.components)) {
    errors.push("components must be a list");
    return errors;
  }

  const claimCardIds = new Set();
  const componentsByClaim = new Map();
  const calibrationSourceByClaim = new Map();
  const modelExplanationClaims = new Set();

  scene.components.forEach((component, index) => {
    if (!component || typeof component !== "object") {
      errors.push(`component[${index}] is not a JSON object`);
      return;
    }
    const { type, id, props } = component;
    if (typeof type !== "string") {
      errors.push(`component[${index}] missing type`);
      return;
    }
    if (typeof id !== "string") {
      errors.push(`component[${index}] missing id`);
    }
    if (!props || typeof props !== "object") {
      errors.push(`component[${index}] (${type}) missing props`);
      return;
    }

    const required = REQUIRED_PROPS_BY_COMPONENT[type];
    if (!required) {
      errors.push(`component[${index}] unknown type "${type}"`);
      return;
    }
    for (const propName of required) {
      if (!(propName in props)) {
        errors.push(`component[${index}] (${type}) missing prop: ${propName}`);
      }
    }

    if (type === COMPONENT_CALIBRATION_BADGE) {
      if (!CALIBRATION_SOURCES.includes(props.source)) {
        errors.push(
          `component[${index}] CalibrationBadge has invalid source "${props.source}"`,
        );
      }
      if (typeof id === "string" && id.includes(".") && typeof props.source === "string") {
        calibrationSourceByClaim.set(id.slice(id.indexOf(".") + 1), props.source);
      }
    }

    if (type === COMPONENT_CLAIM_CARD) {
      if (typeof props.claim_id === "string") {
        claimCardIds.add(props.claim_id);
      }
    } else if (typeof id === "string" && id.includes(".")) {
      const claimId = id.slice(id.indexOf(".") + 1);
      if (!componentsByClaim.has(claimId)) {
        componentsByClaim.set(claimId, new Set());
      }
      componentsByClaim.get(claimId).add(type);
      if (type === COMPONENT_MODEL_EXPLANATION_PANEL) {
        modelExplanationClaims.add(claimId);
      }
    }
  });

  if (typeof scene.claim_count === "number" && scene.claim_count !== claimCardIds.size) {
    errors.push(
      `claim_count ${scene.claim_count} does not match number of ClaimCards ${claimCardIds.size}`,
    );
  }

  for (const claimId of claimCardIds) {
    const present = componentsByClaim.get(claimId) || new Set();
    for (const componentType of ALWAYS_PRESENT_COMPONENTS) {
      if (componentType === COMPONENT_CLAIM_CARD) {
        continue;
      }
      if (!present.has(componentType)) {
        errors.push(`claim "${claimId}" missing required component: ${componentType}`);
      }
    }
    if (
      modelExplanationClaims.has(claimId)
      && calibrationSourceByClaim.get(claimId) !== "model-adjusted"
    ) {
      errors.push(
        `claim "${claimId}" has ModelExplanationPanel but CalibrationBadge source is not model-adjusted`,
      );
    }
  }

  return errors;
}
